File: c/SparseMatrix.cpp
#include<vector>
#include "SparseMatrix.h"
using namespace std;
#ifndef SPARSE_MATRIX
#define SPARSE_MATRIX
SparseMatrix::SparseMatrix(){
}
map<int,double> SparseMatrix::pickInstanceFromIsolated(stack <int> &isolated){
    if (!isolated.empty()){
      int idx = isolated.top();
      isolated.pop();
      return getInstanceFromMatrix(idx);
    }else
      return randomlyPickAnInstance();
}
void SparseMatrix::updateCentroid(int instanceIndex, map<int,double> &centroid){
    int featureIndex_1 = getFeaturesByInstance[instanceIndex].first;
    int featureIndex_2 = getFeaturesByInstance[instanceIndex].second;
    centroid[featureIndex_1] = ( centroid.find(featureIndex_1) != centroid.end() )?centroid[featureIndex_1] + 1:1;
    centroid[featureIndex_2] = ( centroid.find(featureIndex_2) != centroid.end() )?centroid[featureIndex_2] + 1:1;
}
map<int,double> SparseMatrix::randomlyPickAnInstance(){
    int instanceIndex = rand()%n_size;
    map<int,double> instance = getInstanceFromMatrix(instanceIndex);
    return instance;
}
map<int,double> SparseMatrix::getInstanceFromMatrix(int instanceIndex){
    map<int,double> instance;
    int featureIndex_1 = getFeaturesByInstance[instanceIndex].first;
    int featureIndex_2 = getFeaturesByInstance[instanceIndex].second;
    instance[featureIndex_1] = 1;
    instance[featureIndex_2] = 1;
    return instance;
}
set<int> SparseMatrix::getRelevantInstanceSetByFeatureIndexAndCentroid(map<int,double> &centroid){
    set<int> relevantInstanceSet;
    int featureIndex, instanceIndex;
    for(map<int,double>::iterator it = centroid.begin();  it != centroid.end(); it++){
        featureIndex = it->first;
        for(int i=0;i<relevantInstance[featureIndex].size();i++){
            instanceIndex = relevantInstance[featureIndex][i];
            relevantInstanceSet.insert(instanceIndex);
        }
    }
    return relevantInstanceSet;
}
double SparseMatrix::calculateSimilarity(int instanceIndex, map<int,double> &centroid){
    double similarity = 0;
    int featureIndex_1 = getFeaturesByInstance[instanceIndex].first;
    int featureIndex_2 = getFeaturesByInstance[instanceIndex].second;
    similarity += ( centroid.find(featureIndex_1) != centroid.end() )?centroid[featureIndex_1]:0;
    similarity += ( centroid.find(featureIndex_2) != centroid.end() )?centroid[featureIndex_2]:0;
    return similarity;
}
bool SparseMatrix::loadFile(const char* filename){
    FILE *fp = fopen(filename, "r");
    if (fp) {
      int node_1,node_2;
      m_size = n_size = 0;
      getFeaturesByInstance.clear();
      relevantInstance.clear();
      // build m -> nodes n -> edges matrix
      // m = nodes, n = edges_size
      while (fscanf(fp, "%d,%d", &node_1,&node_2) !=EOF ){
        m_size = (node_1>m_size)?node_1:m_size;
        m_size = (node_2>m_size)?node_2:m_size;
        getFeaturesByInstance.push_back(make_pair(node_1,node_2));
        n_size++;
      }

      // build feature to instance
      for(int i=0;i<m_size;i++)
        relevantInstance.push_back(vector<int>());
      for(int instanceIndex=0;instanceIndex<n_size;instanceIndex++){
        int featureIndex_1 = getFeaturesByInstance[instanceIndex].first;
        int featureIndex_2 = getFeaturesByInstance[instanceIndex].second;
        relevantInstance[featureIndex_1].push_back(instanceIndex);
        relevantInstance[featureIndex_2].push_back(instanceIndex);
      }
      fclose(fp);
      return true;
    }
    return false;
}
int SparseMatrix::getNodeSize(){
  return m_size;
}
int SparseMatrix::getInstanceSize(){
  return n_size;
}
#endif

File: python/SparseMatrix.py
import random
import scipy.io as sio

class SparseMatrix():
    def __init__(self, fileName = 'blogcatalog.mat'):
        self._data = None
        self.edge_size = 0
        self.node_size = 0
        self.InstanceToFeature = list()
        self.FeatureToInstance = list()
        self.load_from_mat(fileName)
    def load_from_mat(self,data):
        # construct a mapping from features to instances
        mat_contents = sio.loadmat(data)
        network = mat_contents['network']
        self.feature_size = network.shape[0]
        self.instance_size = network.nnz // 2
        rows,columns = network.nonzero()
        network_size = network.nnz
        count_element = 0
        count_row = 0
        for i in range(self.instance_size):
            self.InstanceToFeature.append(list())
        for featureIndex in range(self.feature_size):
            self.FeatureToInstance.append(list())
        for i in range(network_size):
            node_1 = rows[i]
            node_2 = columns[i]
            if node_2 > node_1:
                # instnace to feature
                self.InstanceToFeature[count_row].append(node_1)
                self.InstanceToFeature[count_row].append(node_2)
                # feature to instance
                self.FeatureToInstance[node_1].append(count_row)
                self.FeatureToInstance[node_2].append(count_row)
                count_element = count_element + 2
                count_row = count_row +1
    def getRelevantInstanceSetByFeatureIndexAndCentroid(self,centroid):
        relevantInstanceSet = set()
        # centroid contains featureIndex and value
        for featureIndex in centroid:
            for instanceIndex in self.FeatureToInstance[featureIndex]:
                relevantInstanceSet.add(instanceIndex)
        return relevantInstanceSet
    def pickRandomInstance(self):
        instanceIndex = random.randint(0, self.getInstanceSize() - 1)
        instance = self.dataToInstance(instanceIndex)
        return instance
    def pickInstanceFromIsolated(self, isolated):
        instanceIndex = 0
        if isolated:
            instanceIndex = isolated.pop()
            return self.dataToInstance(instanceIndex);
        else:
            return self.pickRandomInstance()
    def dataToInstance(self,instanceIndex):
        instance = {}
        for featureIndex in self.InstanceToFeature[instanceIndex]:
            instance[featureIndex] = 1
        return instance
    def calculateSimilarity(self, instanceIndex, centroid):
        similarity = 0
        for featureIndex in self.InstanceToFeature[instanceIndex]:
            if featureIndex in centroid:
                similarity += centroid[featureIndex]
        return similarity
    def getFeatureSize(self):
        return self.feature_size
    def getInstanceSize(self):
        return self.instance_size
    def getMatrix(self):
        return self._data
